firewood/src/merkle/node.rs

```diff
@@ -315,6 +330,16 @@ impl Node {
     }
 }
 
+pub(crate) fn write_branch<F: FnOnce(&mut Box<BranchNode>)>(f: F) -> impl FnOnce(&mut Node) {
```

Comment:
Unless I'm missing something, this function:
 - is named "write_branch" but does not write anything
 - seems to convert a Node into a Box<BranchNode>
 - wraps some arbitrary function with the mapping logic

This signature appears to be needlessly complicated: all callers but one seem to save then invoke the returned closure. The one that doesn't expects a closure, which would be pretty easy to code.

Maybe a method on `Node` that does the heavy lifting would improve readability (inside `impl Node`):

```
    // Assume this node is a branch, and return a mutable boxed BranchNode
    // Caution: panics if you call it with a different kind of node
    pub(crate) fn as_branch_mut(&mut self) -> &mut Box<BranchNode> {
        self.inner_mut().as_branch_mut().expect("must be a branch node")
    }
```
This changes this call sequence, repeated a few times in the code:
```
let f = write_branch(|branch| {
    branch.children[next_nibble as usize] = Some(new_leaf);
});
f(node)
```
Into:
```
node.as_branch_mut().children[next_nibble as usize] = Some(new_leaf);
```